File: typescript/packages/bench/src/reports/orderBacklogReport.ts
/** Open orders grouped by lifecycle state. */

import type { ReportRow } from "../contracts/reportContract";
import type { Dataset } from "../dataset";
import { reportRow } from "../contracts/reportContract";
import { RepairStatus } from "@atelier/core";
import { AbstractReport } from "../support/abstractReport";

/** Open orders grouped by lifecycle state. */
export class OrderBacklogReport extends AbstractReport {
    /** Registry slug; also the URL segment. */
    static readonly SLUG = "order-backlog";

    constructor() {
        super(OrderBacklogReport.SLUG, "Order backlog");
    }

    override rows(data: Dataset): ReportRow[] {
        const counts = new Map<string, number>();
        for (const order of data.openOrders()) {
            const label = RepairStatus.label(order.status);
            counts.set(label, (counts.get(label) ?? 0) + 1);
        }

        return [...counts.entries()]
            .sort(([left], [right]) => left.localeCompare(right))
            .map(([label, count]) => reportRow(label, count));
    }
}

File: rust/crates/atelier-bench/src/reports/order_backlog_report.rs
//! Open orders grouped by lifecycle state.

use crate::contracts::report_contract::{ReportContract, ReportRow};
use crate::dataset::Dataset;
use crate::support::abstract_report::ReportBase;
use itertools::Itertools;

/// Open orders grouped by lifecycle state.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct OrderBacklogReport {
    base: ReportBase,
}

impl OrderBacklogReport {
    /// Registry slug; also the URL segment.
    pub const SLUG: &'static str = "order-backlog";

    pub const fn new() -> Self {
        Self {
            base: ReportBase::new(Self::SLUG, "Order backlog"),
        }
    }

    /// Shared identity and rendering precision.
    pub const fn base(&self) -> &ReportBase {
        &self.base
    }
}

impl Default for OrderBacklogReport {
    fn default() -> Self {
        Self::new()
    }
}

impl ReportContract for OrderBacklogReport {
    fn slug(&self) -> &'static str {
        self.base.slug()
    }

    fn title(&self) -> &'static str {
        self.base.title()
    }

    fn rows(&self, data: &Dataset) -> Vec<ReportRow> {
        data.open_orders()
            .iter()
            .map(|order| order.status.label())
            .counts()
            .into_iter()
            .sorted_by_key(|(label, _)| *label)
            .map(|(label, count)| ReportRow::new(label, count as f64))
            .collect()
    }
}
